### ISSUE TITLE
No links between Dell switches 

### ISSUE THREAD

**Issue by mileswu123** [netdisco/snmp-info#340]
When I tested Netdisco with Dell switches, Netdisco only shows neighbors when the switch's IP address is in lldp_rman_addr {...}

## Expected Behavior

There is link between switch 162.16.0.10/port Gi1/0/30 and switch 162.17.0.206/port Gi1/0/12. I expect Netdisco can derive the link from the LLDP data. 

## Current Behavior

Netdisco only shows neighbors when the switch's IP is in lldp_rman_addr {...} When I looked into SNMP/Info.pm and SNMP/Info/LLDP.pm, I couldn't see how it handle this kind of LLDP data.

## Possible Solution

I used ./netdisco-do show -d 162.16.0.10 -e all to collect the following info (I removed unrelated data)

stp_i_mac               
 {  0   "wõä" }

e_name                   
{  45 "Gi1/0/30", }

------------------------------------------------------------
lldp_rem_id              
{  42849478.42.7183   "‰„Ñπ", }

 lldp_rem_id_type         
{  42849478.42.7183    "macAddress", }  
   
 lldp_rem_media_cap       
{  42849478.42.7183   "\0", }
   
lldp_rem_pid             
{  42849478.42.7183   "Gi1/0/12", }

lldp_rem_pid_type        
{  42849478.42.7183   "interfaceName", }

lldp_rem_sys_cap         
{  42849478.42.7183   "\0\0", }  

lldp_rem_sysname         
{  42849478.42.7183   “2000_Switch", }

lldp_rem_sysdesc         
{  42849478.42.7183   "", } 

And I used ./netdisco-do show -d 162.17.0.206  -e all to collect the following info (I removed unrelated data)

stp_i_mac                
{  0   "‰„Ñπ" }

e_name                  
 {  4    "Gi1/0/12", }

-------------------------------------------------------------
lldp_rem_cap_spt        
{  21768813.1.2291    "\0\0",}

lldp_rem_desc            
{  21768813.1.2291    “Rm2000", }
    
lldp_rem_id              
{  21768813.1.2291    "wõä", }

lldp_rem_id_type         
{  21768813.1.2291   "macAddress", }

lldp_rem_media_cap       
{  21768813.1.2291   "\0", }

lldp_rem_pid             
{ 21768813.1.2291    "Gi1/0/30", }

lldp_rem_pid_type        
{ 21768813.1.2291   "interfaceName", }

lldp_rem_sys_cap        
{  21768813.1.2291    "\0\0", }

lldp_rem_sysdesc         
{ 21768813.1.2291    “2ndFloor_Switch", }
 
lldp_rem_sysname         
{ 21768813.1.2291    “", }
    
By using the LLDP data with macAddress type as the id. Netdisco can search the id from the stp_i_mac of all the switches to locate the neighbored switch and also linked ports between the two switches.    

## Steps to Reproduce (for bugs)

1. 
2. 
3. 
4. 

## Your Environment

* Netdisco version used: 2.40.6
* SNMP::Info version used: 3.65

**Comment by mileswu123** [comment]
I just replied to Nick email. Please let me know if you didn't receive it.

**Comment by inphobia** [comment]
at first glance i haven't gotten anything yet, nor did i see anything blocked in the spamfilter. it's: nick.nauwelaerts  @  aquafin.be

**Comment by mileswu123** [comment]
Did you receive it? I sent email to nick.nauwelaerts@aquafin.be on 5/28.

**Comment by inphobia** [comment]
got it, but still figuring out the issue.

while dumping mac addresses in lldp_ip might work for you i'm not fond of that solution. from my current understanding netdisco will fail to save relations if it does not know the ip (ipv4 or 6) of the remote device:

https://github.com/netdisco/netdisco/blob/0f1c264e410c1ee4de7c44a8f7550356c09225dd/lib/App/Netdisco/Worker/Plugin/Discover/Neighbors.pm#L130-L140

@ollyg :
do we support topo stuff when the remote device has no easily determined ip address?

**Comment by mileswu123** [comment]
Looks like Dell switch returns stp_i_mac in lldp_rem_id query, if lldp_rem_id_type is mac address.  I think that the following changes could make this work.

1. Query and store stp_i_mac into DB device table during the device discovery.
2. Modify the function sub lldp_ip {} in LLDP.pm  
sub lldp_ip {
    my $lldp    = shift;
    my $partial = shift;

    my $rman_addr = $lldp->lldp_rman_addr($partial) || {};

    my %lldp_ip;
    foreach my $key ( keys %$rman_addr ) {
        my ( $index, $proto, $addr ) = $lldp->_lldp_addr_index($key);
        next unless defined $index;
        next unless $proto == 1;
        $lldp_ip{$index} = $addr;     
    }

    # Look up switch IP address by stp_i_mac
    my $ch_type = $lldp->lldp_rem_id_type($partial) || {};
    my $ch      = $lldp->lldp_rem_id($partial)      || {};

    foreach my $key ( keys %$ch ) {
        my $id = $ch->{$key};
        next unless $id;
        my $type = $ch_type->{$key};
        next unless !defined  $lldp_ip{$key};

        if ( $type =~ /mac/ ) {
            # someone needs to implement function lookup_ip_by_stp_i_mac()
            my $ip = SNMP::Info::lookup_ip_by_stp_i_mac($id)
            if ($ip)
                 $lldp_ip{$key} = $id;
        }
    }
    return \%lldp_ip;
}

**Comment by inphobia** [comment]
tagging @JeroenvIS after i saw him discussing this on irc

**Comment by inphobia** [comment]
>     1. Query and store stp_i_mac into DB device table during the device discovery.

sounds like a good idea, the problem is this will not always be unique. (at least in cisco nexus vpc setups this will be shared between vtp members). on the other hand, we alrdy have a mac field in the device table, which comes from the mac() function, but not all snmp::info classes support this, nor is it the guaranteed to be the same as stp_i_mac. (stp_i_mac is not advertised in the docs, the b_mac from s::i::bridge should take care of that.)

that all being said, i am a fan of adding the stp base adress to device info, either as the column "mac" in the device table, or as a new column.

after reviewing your logs & the code again, all the lldp mappings are done with lldpRemManAddrIfSubtype (the lldp_rman_addr function in s::i::lldap) or with lldpLocPortDesc.

now, the first switch only returns 1 entry here:
```
[4969] 2019-05-07 00:20:45  info show: [162.16.0.206]/lldp_rman_addr started at Tue May  7 00:20:45 2019
\ {
    15671549.52.1464.1.4.162.16.4.189   "ifIndex"
}
```

that's most likely the reason why it's missing so many neighbors.

i went through all the release notes for n1100 & n1500 switches, and noticed 2 bugs:

![image](https://user-images.githubusercontent.com/39005454/59138301-9503ed00-898c-11e9-891b-71fde61bad7b.png)
![image](https://user-images.githubusercontent.com/39005454/59138314-a3520900-898c-11e9-80a9-be60f5953893.png)

i'm no fan of having lldp_ip() return anything other as an ip address, that's not it's function. lldp_addr() should return an address if it's known for all devices, but as seen in your logs it only returns 1 peer. since your switches are running different os versions i wonder if that might could be part of the issue (wrt those bugs).

ofcourse, none of that fixes your issue :)


perhaps an approach would be to store lldpLocChassisId or anything else from the lldpLoc tree for each device and try to match against that as a last resort.

**Comment by mileswu123** [comment]
For experiment, I got it working by reusing the mac field in Device table, modifying Layer2.pm and Layer3.pm to retrieve the stp_i_map and store them into DB. Then I modified lldp_ip() in LLDP.pm to map the lldp_rem_id to ip, when lldp_rem_id_type is Mac address. Please see the attached file.

After rediscovering all devices, which will get stp_i_mac for all layer2/3 devices and find the neighbor through lldp_ip{}, I got all three missing linkages.

WHERE me.ip = '172.17.0.10' AND me.port = 'Gi1/0/16'
[10577] 2019-06-10 23:13:57 debug  [172.17.0.10] neigh - 172.17.1.207 with ID [e4:f0:04:e3:84:b9] on Gi1/0/42

WHERE me.ip = '172.17.0.10' AND me.port = 'Gi2/0/20'
[10577] 2019-06-10 23:13:57 debug  [172.17.0.10] neigh - 172.17.1.206 with ID [4c:d9:8f:e0:d6:f5] on Gi1/0/31

WHERE me.ip = '172.17.0.10' AND me.port = 'Gi1/0/2'
[10577] 2019-06-10 23:13:57 debug  [172.17.0.10] neigh - 172.17.0.1 with ID [20:04:0f:18:60:f2] on Te1/0/3

These Dell switches are running with 6.2.6.6, 6.4.1.4 and 6.5.2.23 firmware. 

[Archive.zip](https://github.com/netdisco/snmp-info/files/3274095/Archive.zip)

**Comment by inphobia** [comment]
i haven't forgotten about this but have been wondering what the best approach would be to tac;kle this problem.

i'm thinking this could be a multistep approach:
1. netdisco uses the s::i::mac() function to get the input it uses to fill out the mac address field in the device info tab. this function in only available on s::i::layer3.pm and several device specific modules. netdisco should try harder to get a useable address here, but i'm not sure which is the better way:
   * move mac() from layer3 to info.pm so it becomes a basic function
     * side note; most, but not all, network devices use mac addresses, it's mostly for ethernet & etherlike stuff. other protocols have different means for linklocal addressing.
    * perhaps netdisco needs some kind of system id field instead of a mac field.
    * or perhaps we need to start saving multiple device ids (mac, stp root, lldp id, etc...)
   * have a mac() function in layer2 like the change you proposed. i don't however agree with the layer3.pm change. dot1dBaseBridgeAddress only belongs in layer2 imo. either netdisco or snmp::info main should prefer one over the other.
1. it seems netdisco currently might has issues c_*() functions returning multiple matching entries for the same device, my oppinion is that snmp::info does the correct thing by returning all permutations it knows about, and then let netdisco device in which priority to try & resolve those relations.
1. finally, it seems those delll devices might return s::i::interfaces() results index on user defined names for those interfaces. while this is perfectly legal from an snmp standpoint this might confuse netdisco if those user assigned names change (i have a supsiscion this is the case but have yet to dive deepe into it. so:
  * perhaps snmp::info should have some kind of guideline what it returns as interfaces() result.
    * either define an order in which datasources can be used
    * or somehow enforce that each module tries and return the most stable interface identifier as interfaces() result.
       * (that's the reason the lenovo module does this:)
https://github.com/netdisco/snmp-info/blob/02c8ed0764dc2230ea1cd31028f63de655085cc5/lib/SNMP/Info/Layer3/Lenovo.pm#L72-L79
1. your lldp.pm changes are a working approach to solving the last piece of the puzzle, but snmp::info is made to query devices. while it's almost exclusivly used by netdisco we can't use netdisco functions in snmp::info regretfully. i'm still of the opinion snmp::info is doing it's job and we need a bit more logic in netdisco to get this working.

now, most of these changes might have a bigger impact as expected, i'll open issues for the individual steps to resolve and to request de feedback on what's the preferred way to handle this.

bottom line: it's on our radar but will take time to get everything into place to make thiss work.

tagging @netdisco/snmp-info-developers  & @netdisco/snmp-info-admins  here for more feedback on best approach.

**Comment by inphobia** [comment]
https://github.com/netdisco/snmp-info/issues/342#issuecomment-502444168 has a testscript to collect all known relations. is there any chance you could run that on both devices?

that would give me a better idea on how to refactor netdisco, since imo snmp::info is doing this correctly, but netdisco insists on having a remote ip (for which there are valid reasons, since looking for ips is a lot quicker as checking each know port mac), but perhaps there are ways around this.

thx

**Comment by edo1** [comment]
Any news? IMO this is an important thing, some devices announce MAC address only.

**Comment by inphobia** [comment]
after an hiatus i see #395 also has the same issues. the reason this issue is still open is because it's still a problem, regretfully also one without an easy fix.
we have not forgotten this, nor do we have a solid solution right now. it's taking longer that any of us wanted, but we're working on it.

**Comment by inphobia** [comment]
netdisco/netdisco#737 also feels very similar.

**Comment by JeroenvIS** [comment]
> Any news? IMO this is an important thing, some devices announce MAC address only.

IMHO it depends on how devices advertise themselves.

First of all, there is the Chassis ID that is advertised: _"The string value used to identify the chassis component associated with the remote system."_ The device will also say what type of ID it is sending, the lldpRemChassisIdSubtype. There are 7 defined types: _"{chassisComponent(1), interfaceAlias(2), portComponent(3), macAddress(4), networkAddress(5), interfaceName(6), local(7)}"_. The chassis ID and type are always present in LLDP neighbor info.

Then there is the remote management address(es) that can be advertised. A device can advertise multiple management addresses of different types. _"The string value used to identify the management address component associated with the remote system. The purpose of this address is to contact the management entity."_. The remote management address info is optional to send; what I see in our network is that generally the network components all advertise at least one management address, as do some IP phones, but other IP phones and some servers (eg ESX) and other clients just run LLDP and don't advertise any management info.

Currently, to determine neighbour relations, Netdisco only looks at management addresses that are advertised. IMHO that's a good choice, because we only store relations for devices that are known to Netdisco, and Netdisco needs a management address to connect to anyway.

**Personally I'd first try to investigate why some devices in the management domain do not advertise a management address to their neighbours, but only send their chassis identifier.**

If it's really not possible to configure these devices to send a management address in their LLDP messages, then we could consider extending Netdisco to also look at Chassis ID to determine potential neighbour relations. Of course that logic should then be able to deal with the different types of chassis identifiers that may be reported, or at least not break if the remote ID is not a MAC address but one of the other 6 types.

**Comment by JeroenvIS** [comment]
Result of a quick search on the Internet: looks like various Dell models don't send remote management info through LLDP by default, but can be configured to do so. I strongly suggest Dell users to investigate that angle before we try extensive changes to Netdisco and SNMP::Info, which could have unexpected side effects and easily introduce new bugs in determining neighbour relations.

**Comment by edo1** [comment]
@JeroenvIS 
Any reason to do **not** use MAC addresses to determine neighbor relations?

> Of course that logic should then be able to deal with the different types of chassis identifiers that may be reported, or at least not break if the remote ID is not a MAC address but one of the other 6 types.

Ignoring all types except a MAC address doesn't seem difficult.

**Comment by inphobia** [comment]
> @JeroenvIS
> Any reason to do **not** use MAC addresses to determine neighbor relations?

yeah, code complexity foremost. if all we had was lldp perhaps we could add some cludges to use different identification models. to give you an idea, this is the lldp mib: http://www.circitor.fr/Mibs/Mib/L/LLDP-MIB.mib
oh, i forgot to mention lldp-ext-med-mib, lldp-ext-dot1-mib & lldp-ext-dot3-mib

and then we also have cdp, edp, fdp & whatever

all updates to either our code or documentation (https://metacpan.org/pod/SNMP::Info::LLDP) are welcome, but perhaps @JeroenvIS  could link the required config lines to make dell switches send their management info so we can add that to our wiki?

imo, discovery fixup is the second most common code in snmp::info after just plain broken snmp "support".

**Comment by kinsonchan** [comment]
Hello,

Thank you for the software and the discussion.  May I be shed some light how I can change a Dell switch from advertising its MAC address as chassis ID to the management IP address, for the above purpose of switch matching?  I tried options like `lldp management_adr_tlv ipv4 virtual_ip` and also configuring the IP address of `mgmt 1/1/1` without much avail...

Thanks,
Kinson
